File: AOJ/ITP1/011/D_Dice4.cpp
#include <iostream>
#include <algorithm>
#include <iterator>
using namespace std;

struct Dice
{
  int top;
  int bottom;
  int north;
  int east;
  int south;
  int west;
  void toNorth()
  {
    int t = top, b = bottom, n = north, s = south;
    top = s, bottom = n, north = t, south = b;
  }
  void toEast()
  {
    int t = top, b = bottom, e = east, w = west;
    top = w, bottom = e, east = t, west = b;
  }
  void toSouth()
  {
    int t = top, b = bottom, n = north, s = south;
    top = n, bottom = s, north = b, south = t;
  }
  void toWest()
  {
    int t = top, b = bottom, e = east, w = west;
    top = e, bottom = w, west = t, east = b;
  }
  void toTop(int a)
  {
    if (a == north)
      toSouth();
    else if (a == east)
      toWest();
    else if (a == south)
      toNorth();
    else if (a == west)
      toEast();
    else if (a == bottom)
    {
      toNorth();
      toNorth();
    }
  }
  void spin()
  {
    int n = north, e = east, s = south, w = west;
    north = w, east = n, south = e, west = s;
  }
  bool equals(Dice d)
  {
    int n1[6] = {top, bottom, north, west, south, east};
    int n2[6] = {d.top, d.bottom, d.north, d.west, d.south, d.east};
    bool matchNumbers = true;
    for (int i = 0; i < 6 && matchNumbers; i++)
      matchNumbers = find(begin(n2), end(n2), n1[i]) != end(n2);
    if (!matchNumbers)
      return false;

    toTop(d.top);
    int count = 0;
    while (south != d.south && count <= 3)
    {
      spin();
      count++;
    }

    if (top == d.top && bottom == d.bottom &&
        north == d.north && east == d.east &&
        south == d.south && west == d.west)
      return true;
    else
      return false;
  }
};

int main()
{
  int n;
  cin >> n;
  Dice dices[n];
  for (int i = 0; i < n; i++)
  {
    dices[i] = Dice{};
    cin >> dices[i].top >> dices[i].south >> dices[i].east;
    cin >> dices[i].west >> dices[i].north >> dices[i].bottom;
  }

  bool notEqual = true;
  for (int i = 0; i < n && notEqual; i++)
    for (int j = i + 1; j < n && notEqual; j++)
      notEqual = !dices[i].equals(dices[j]);

  if (notEqual)
    cout << "Yes" << endl;
  else
    cout << "No" << endl;

  return 0;
}

File: AOJ/ITP1/011/D_Dice4.php
<?php

class Dice
{
  public $top = 0;
  public $bottom = 0;
  public $north = 0;
  public $east = 0;
  public $south = 0;
  public $west = 0;

  public function toNorth()
  {
    $t = $this->top;
    $this->top = $this->south;
    $this->south = $this->bottom;
    $this->bottom = $this->north;
    $this->north = $t;
  }

  public function toEast()
  {
    $t = $this->top;
    $this->top = $this->west;
    $this->west = $this->bottom;
    $this->bottom = $this->east;
    $this->east = $t;
  }

  public function toSouth()
  {
    $t = $this->top;
    $this->top = $this->north;
    $this->north = $this->bottom;
    $this->bottom = $this->south;
    $this->south = $t;
  }

  public function toWest()
  {
    $t = $this->top;
    $this->top = $this->east;
    $this->east = $this->bottom;
    $this->bottom = $this->west;
    $this->west = $t;
  }

  public function spin()
  {
    $n = $this->north;
    $this->north = $this->west;
    $this->west = $this->south;
    $this->south = $this->east;
    $this->east = $n;
  }

  public function toBottom()
  {
    $this->toNorth();
    $this->toNorth();
  }

  public function toTop($x)
  {
    if ($x == $this->north) $this->toSouth();
    elseif ($x == $this->east) $this->toWest();
    elseif ($x == $this->south) $this->toNorth();
    elseif ($x == $this->west) $this->toEast();
    elseif ($x == $this->bottom) $this->toBottom();
  }

  public function equals($d)
  {
    $n1 = array($this->top, $this->bottom, $this->north, $this->east, $this->south, $this->west);
    $n2 = array($d->top, $d->bottom, $d->north, $d->east, $d->south, $d->west);
    sort($n1);
    sort($n2);
    if ($n1 != $n2) return false;

    $this->toTop($d->top);
    $count = 0;
    while ($this->south != $d->south && $count <= 3) {
      $this->spin();
      $count++;
    }

    if ($this->top == $d->top && $this->bottom == $d->bottom &&
      $this->north == $d->north && $this->east == $d->east &&
      $this->south == $d->south && $this->west == $d->west)
      return true;
    else
      return false;
  }
}

$n = trim(fgets(STDIN));
$dices = array();
for ($i = 0; $i < $n; $i++) {
  $num = explode(" ", trim(fgets(STDIN)));
  array_push($dices, getDice($num));
}

$notEqual = true;
for ($i = 0; $i < $n && $notEqual; $i++)
  for ($j = $i + 1; $j < $n && $notEqual; $j++)
    $notEqual = !$dices[$i]->equals($dices[$j]);

if ($notEqual) echo "Yes\n";
else echo "No\n";

function getDice($n)
{
  $dice = new Dice();
  $dice->top = $n[0];
  $dice->south = $n[1];
  $dice->east = $n[2];
  $dice->west = $n[3];
  $dice->north = $n[4];
  $dice->bottom = $n[5];
  return $dice;
}
